browser/ui/views/playlist/playlist_bubbles_controller.h

```diff
@@ -0,0 +1,56 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_BROWSER_UI_VIEWS_PLAYLIST_PLAYLIST_BUBBLES_CONTROLLER_H_
+#define BRAVE_BROWSER_UI_VIEWS_PLAYLIST_PLAYLIST_BUBBLES_CONTROLLER_H_
+
+#include "base/memory/raw_ptr.h"
+#include "base/memory/weak_ptr.h"
+#include "content/public/browser/web_contents_user_data.h"
+
+class PlaylistActionIconView;
+
+namespace content {
+class WebContents;
+}
+
+namespace playlist {
+class PlaylistBubble;
+
+class PlaylistBubblesController
+    : public content::WebContentsUserData<PlaylistBubblesController> {
+ public:
+  enum class BubbleType { kInfer, kAdd, kConfirm };
+
+  static PlaylistBubblesController* CreateOrGetFromWebContents(
+      content::WebContents* web_contents);
+
+  ~PlaylistBubblesController() override;
+
+  void ShowBubble(base::WeakPtr<PlaylistActionIconView> anchor_view,
+                  BubbleType bubble_type = BubbleType::kInfer);
+
+  PlaylistBubble* GetBubble();
+
+  void OnBubbleClosed();
+
+  base::WeakPtr<PlaylistBubblesController> AsWeakPtr();
+
+ protected:
+  explicit PlaylistBubblesController(content::WebContents* web_contents);
+
+ private:
+  friend class content::WebContentsUserData<PlaylistBubblesController>;
+
+  raw_ptr<PlaylistBubble> bubble_ = nullptr;
+
+  base::WeakPtrFactory<PlaylistBubblesController> weak_ptr_factory_{this};
```

Comment:
ditto: Moving weak_ptr_factory to the last would be better.

Reply:
I haven't seen a Chromium guideline on the order here, so I usually go with what's prevalent in Chromium:
1.  `base::WeakPtrFactory<>`
2. `WEB_CONTENTS_USER_DATA_KEY_DECL()`

This also doesn't go against the `base::WeakPtrFactory<>` guidelines, as `WEB_CONTENTS_USER_DATA_KEY_DECL()` declares a `static` variable (which has no effect on the order in which members are destructed).

Reply:
Actually, as you noted `WEB_CONTENTS_USER_DATA_KEY_DECL()` is `static` variable, which made me torn. As per the Google's guideline, static variable should be first and then non-static variable. Then, WEB_CONTENTS_USER_DATA_KEY_DECL() should be moved up. But you know, put this at the end of class is prevalent.

Reply:
I'm okay with the current state but just want to hear what you think about it.

Reply:
Wondering why they don't have a Clang tool built into presubmit to auto-correct these (`WEB_CONTENTS_USER_DATA_KEY_DECL()`, or even `base::WeakPtrFactory<>`). It's easily doable... leaning towards the `static` point, so let's move it up. 👍🏻